nixos/modules/services/web-apps/immich.nix

```diff
@@ -0,0 +1,246 @@
+        '';
+      };
+    };
+
+    database = {
+      enable = mkEnableOption "postgresql" // { default = true; };
+      createDB = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to create the database for immich automatically.";
+      };
+      name = mkOption {
+        type = types.str;
+        default = "immich";
+        description = "The name of the immich database.";
+      };
+      host = mkOption {
+        type = types.str;
+        default = "/run/postgresql";
+        example = "127.0.0.1";
+        description = "Hostname or address of the postgresql server. If an absolute path is given here, it will be interpreted as a unix socket path.";
+      };
+      user = mkOption {
+        type = types.str;
+        default = "immich";
+        description = "The database user for immich.";
+      };
+      setupPgvectors = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to add the pgvecto.rs extension (required to run immich) to postgresql.";
+      };
+    };
+    redis = {
+      enable = mkEnableOption "immich redis server" // { default = true; };
+      host = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "The host that redis will listen on.";
+      };
+      serverName = mkOption {
+        type = types.str;
+        default = "immich";
+        description = "The server name of the immich redis instance.";
+      };
+      port = mkOption {
+        type = types.port;
+        default = 6379;
+        description = "The port that redis will listen on.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [ {
+      assertion = !isUnixSocket -> cfg.secretsFile != null;
+      message = "A secrets file containing at least the database password must be provided when unix sockets are not used.";
+    } ];
+
+    services.postgresql = mkIf cfg.database.enable {
+      enable = true;
+      ensureDatabases = mkIf cfg.database.createDB [ cfg.database.name ];
+      ensureUsers = mkIf cfg.database.createDB [ {
+        name = cfg.database.user;
+        ensureDBOwnership = true;
+        ensureClauses.login = true;
+      } ];
+      extraPlugins = mkIf cfg.database.setupPgvectors (ps: with ps; [ pgvecto-rs ]);
+      settings = mkIf cfg.database.setupPgvectors {
+        shared_preload_libraries = [ "vectors.so" ];
+        search_path = "\"$user\", public, vectors";
+      };
+    };
+    systemd.services.postgresql.serviceConfig.ExecStartPost = mkIf cfg.database.setupPgvectors [ ''
+      +${pkgs.bash}/bin/bash -c '${pkgs.sudo}/bin/sudo -i -u postgres psql -d "${cfg.user}" -c "CREATE EXTENSION IF NOT EXISTS vectors; CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE; ALTER SCHEMA vectors OWNER TO ${cfg.database.user}; GRANT SELECT ON TABLE pg_vector_index_stat TO ${cfg.database.user};"'
```

Comment:
Is there a way we could use writeTextFile for this?

Reply:
I don't think so, after all this added to an otherwise existing systemd service file

Reply:
Well, at least something like this:
```nix
  systemd.services.postgresql.serviceConfig.ExecStartPost = let
    sqlFile = pkgs.writeTextFile "immich-pgvectors-setup.sql" ''
      CREATE EXTENSION IF NOT EXISTS vectors;
      CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
      ALTER SCHEMA vectors OWNER TO ${cfg.database.user};
      GRANT SELECT ON TABLE pg_vector_index_stat TO ${cfg.database.user};
    '';
  in mkIf cfg.database.setupPgvectors [ ''
    +${pkgs.bash}/bin/bash -c '${pkgs.sudo}/bin/sudo -i -u postgres psql -d "${cfg.user}" -f "${sqlFile}"'
  '' ];
```
should be possible, no?

Also, is the `bash` in front really required? Wouldn't 
```nix
+${pkgs.sudo}/bin/sudo -u postgres psql -d "${cfg.user}" -f "${sqlFile}"
```
not also just work?

Reply:
I also don't see why we need `sudo` here. The postgresql.service is already running as that user.